### ISSUE TITLE
Jupyter Lab stalls when choosing "Restart Kernel and Run All Cells..."

### ISSUE THREAD

**Issue by DevinRouth** [jupyterlab/jupyterlab#9008]
## Description

I recently needed to do a full reinstall of Jupyter Lab on MacOS 10.15. I use Homebrew as my package manager, running Python 3.8.5.

Jupyter Lab installed correctly, as did the R kernel (which I use frequently). However, I noticed that whenever I choose the "Restart Kernel and Run All Cells..." option, the individual notebook within Jupyter Lab stalls until I restart the kernel. Manually selecting and running all cells runs the code without any errors or stalling.

## Reproduce

I tried running the same notebook file by launching a classic Jupyter Notebook from Jupyter Lab, and the "Restart Kernel and Run All Cells..." option worked perfectly.

## Expected behavior

I expected Jupyter Lab to restart the kernel and run all cells without any lag.

## Context

MacOS 10.15.2
Python 3.8.5 (installed via HomeBrew)
Chrome Version 84.0.4147.135

Output of `jupyter --version`:
```
jupyter core     : 4.6.3
jupyter-notebook : 6.1.4
qtconsole        : 4.7.7
ipython          : 7.18.1
ipykernel        : 5.3.4
jupyter client   : 6.1.7
jupyter lab      : 2.2.8
nbconvert        : 6.0.2
ipywidgets       : 7.5.1
nbformat         : 5.0.7
traitlets        : 5.0.4
```

When I run `jupyter lab --debug` and attempt the "Restart Kernel and Run All Cells...", the output continuously prints:

`Accepting token-authenticated connection from ::1` and a variation of either `200 GET /api/sessions?1600160701879 (::1) 1.71ms`, `200 GET /api/kernels?1600160731926 (::1) 0.89ms`, or `200 GET /api/terminals?1600160759346 (::1) 1.16ms`.

I attempted to solve the issue by running `pip3 install --upgrade ipykernel` per [this issue](https://github.com/jupyter/jupyter_console/issues/163#issuecomment-418392676), but it did not solve the problem.

Any help would be appreciated, and thanks!

**Comment by nighcoder** [comment]
I'm having the same issue here.
Output of `jupyter --version`:
```
jupyter core     : 4.6.3
jupyter-notebook : 6.1.4
qtconsole        : 4.7.5
ipython          : 7.18.1
ipykernel        : 5.3.4
jupyter client   : 6.1.7
jupyter lab      : 2.2.7
nbconvert        : 6.0.2
ipywidgets       : 7.5.1
nbformat         : 5.0.7
traitlets        : 5.0.4
```

I'm having this issue most reliably with the clojupyter (clojure) kernel, but sometime I can get the same problem with the python kernel.
I did some debugging on this problem and found that jupyter lab actually restarts the kernel, but never follows up with sending the cells for execution. The cells are marked for execution, but the kernel status is actually idle (you can see it in the lower left corner). You can also check that the kernel is available and is not blocking on anything by just manually sending a cell to execution after you restart the kernel. @DevinRouth , can you reproduce the issues I'm describing?

Note: Running restart kernel from the **Kernel** menu and Run all cells from the **Run** menu work as intented.

**Comment by DevinRouth** [comment]
Yes, I confirm the same behavior as @nighcoder. All cells are run when specified to run from the "Run" menu, but not from the "Kernel" menu. When selecting "Restart Kernel and Run All Cells", the cells are marked with an asterisk and the kernel remains idle.

**Comment by phish108** [comment]
I confirm this behaviour for JL 2.2.8, too. As reported "shutting down" the kernel forcefully from the `running kernels and terminals`-sidetab and then restarting it from the menu works. But restarting the Kernel works from neither menu option in `Run` or `Kernel`.

**Comment by sdegrace** [comment]
```
>jupyter --version
jupyter core     : 4.6.3
jupyter-notebook : 6.1.4
qtconsole        : 4.7.7
ipython          : 7.18.1
ipykernel        : 5.3.4
jupyter client   : 6.1.7
jupyter lab      : 3.0.0rc4
nbconvert        : 6.0.7
ipywidgets       : 7.5.1
nbformat         : 5.0.7
traitlets        : 5.0.4
```
Same issue, though to be clear it doesn't hang the kernel - it just appears to. At any point I can select Run -> Run all Cells and it works as expected.  It seems as if it gets as far as restarting the kernel and preparing the UI for a run all cells operation, then just forgets to do it.

**Comment by phish108** [comment]
@karlaspuldaro Note sure, if this issue is fixed, but I think it should be backported as mentioned in #9160.

**Comment by karlaspuldaro** [comment]
@phish108 2.2.9 release has been already published. I don't see a PR for this issue but once it's fixed, the backport can go in another patch release.

**Comment by jamesmyatt** [comment]
I also have this problem using IRKernel, but not IPython: https://github.com/IRkernel/IRkernel/issues/671

It's not fixed by Jupyterlab 2.2.9: it now stops  with status "R | Idle" and `[*]` next to all cells rather than "R | Restarting". I can confirm that the problem is with "Restart and run all" specifically, because "Restart kernel" then "Run All" separately is OK.

**Comment by SylvainCorlay** [comment]
We are looking into this now with @JohanMabille @jtpio and @martinRenou.

**Comment by jtpio** [comment]
xref https://github.com/jupyter/jupyter_client/issues/593 for more info

**Comment by minrk** [comment]
In investigating jupyter/jupyter_client#593 I'm fairly confident that there is something amiss in the restart & run all logic in JLab 3.0rc13. On restart & run all, after the websocket connection is re-established, JupyterLab sends no messages at all. A manual click restart, then click run all works just fine. It seems like it's waiting for a condition that's never met before starting the "& run all" part.

Adding debug logging to the zmq handler shows that jupyterlab is not sending the requests after restart at all. So the server-side pub/sub issue we are working on over there doesn't seem to be the *only* issue, at least.

<details>
<summary>pip freeze</summary>

```
anyio==2.0.2
appnope==0.1.2
argon2-cffi==20.1.0
async-generator==1.10
attrs==20.3.0
Babel==2.9.0
backcall==0.2.0
bleach==3.2.1
certifi==2020.12.5
cffi==1.14.4
chardet==3.0.4
click==7.1.2
decorator==4.4.2
defusedxml==0.7.0rc1
entrypoints==0.3
idna==2.10
ipykernel==5.4.2
ipython==7.19.0
ipython-genutils==0.2.0
jedi==0.17.2
Jinja2==3.0.0a1
json5==0.9.5
jsonschema==3.2.0
jupyter-client==6.1.7
jupyter-core==4.7.0
-e git+git@github.com:jupyter/jupyter_server.git@a3a3a46b907cabd7be66c639cf161aafbf20b2e5#egg=jupyter_server
jupyterlab==3.0.0rc13
jupyterlab-pygments==0.1.2
jupyterlab-server==2.0.0rc8
MarkupSafe==2.0.0a1
mistune==0.8.4
nbclassic==0.2.5
nbclient==0.5.1
nbconvert==6.0.7
nbformat==5.0.8
nest-asyncio==1.4.3
notebook==6.1.5
packaging==20.8
pandocfilters==1.4.3
parso==0.7.1
pexpect==4.8.0
pickleshare==0.7.5
pip-tools==5.4.0
prometheus-client==0.9.0
prompt-toolkit==3.0.8
ptyprocess==0.6.0
pycparser==2.20
Pygments==2.7.3
pyparsing==3.0.0b1
pyrsistent==0.17.3
python-dateutil==2.8.1
pytz==2020.4
pyzmq==20.0.0
requests==2.25.0
Send2Trash==1.6.0b1
six==1.15.0
sniffio==1.2.0
terminado==0.9.1
testpath==0.4.4
tornado==6.1
traitlets==5.0.5
urllib3==1.26.2
wcwidth==0.2.5
webencodings==0.5.1
```

</details>

**Comment by minrk** [comment]
I tracked this down to the separate tracking of 'restarting' and connectionStatus here: https://github.com/jupyterlab/jupyterlab/blob/b9511a52557d6ae56e597d0f48392993bdc432ae/packages/services/src/kernel/default.ts#L386-L387

`_sendPending()` is triggered when the connection state is back to 'connected' here: https://github.com/jupyterlab/jupyterlab/blob/b9511a52557d6ae56e597d0f48392993bdc432ae/packages/services/src/kernel/default.ts#L1258-L1262

but that does nothing if the `_kernelSession` is still `RESTARTING_KERNEL_SESSION`: https://github.com/jupyterlab/jupyterlab/blob/b9511a52557d6ae56e597d0f48392993bdc432ae/packages/services/src/kernel/default.ts#L1053-L1060

`_kernelSession` is only reset upon receiving a websocket message: https://github.com/jupyterlab/jupyterlab/blob/b9511a52557d6ae56e597d0f48392993bdc432ae/packages/services/src/kernel/default.ts#L567

so I think there are two problems:

1. if connection state resolves to 'connected' before the `_kernelSession` is resolved away from RESTARTING_KERNEL_SESSION, the pending queue is never flushed (it will be flushed on reconnect). It works only if an unprompted iopub message arrives *before* entering the connected state.
2. resolving the `_kernelSession` out of the 'restarting' state assumes an unprompted websocket message will arrive (because all prompting requests are queued if the kernel is restarting), triggering the reset of `_kernelSession`. It is not a safe assumption that this will happen (this related to the root of  jupyter/jupyter_client#593 ). ipykernel sends a `starting` status message, but it is quite possible to miss this while subscriptions are established.


I think this should be fixed with two changes:

1. allow 'restarting' state to resolve after connected state, flushing the queue in both cases
2. if the kernel info requests that are currently registered after restart are *actually sent* when the connection state is resolved, rather than added to `_pendingMessages`, then it all ought to be okay.

I believe the right logic is there to send a kernelInfoRequest on completed connection, but this must be *actually sent*, not added to the pending queue, which is what happens now.

**Comment by phish108** [comment]
@minrk is this the same logic as in 2.2.x releases?

**Comment by minrk** [comment]
I'm not sure, but I don't think so. I have never been able to reproduce the issue with 2.2.9, but I can 100% of the time with 3.0rc13.

**Comment by phish108** [comment]
I can reproduce this issue all the time with 2.2.8 and 2.2.9. (See my message above)

**Comment by minrk** [comment]
#9484 fixes the issue for me with jlab 3.0. I don't know what to say about 2.x, other than I cannot reproduce it except with 3.0

**Comment by jasongrout** [comment]
For reference, the restarting status was added in https://github.com/jupyterlab/jupyterlab/pull/8562
